feature: add CSIVolumeHealth feature and gate

fengzixu opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind feature

#### What this PR does / why we need it:

*why we need it*

According to the change of [kep](https://github.com/kubernetes/enhancements/pull/2286), we decide to move the responsibility (checking volume condition and reporting abnormal volume event to pod ) from external-health-monitor-agent to kubelet.

*what this PR does*

1. Modify [GetMetrics](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/volume.go#L58) function to return the volumeStats objet which includes the volumeCondition info
2. Sending event to the corresponding pod which is using an abnormal volume
3. Make compatibility change on functions which called GetMetrics function 

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->

Not yet.

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Enables Kubelet to check volume condition and log events to corresponding pods.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:

- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/pull/2286
```

fengzixu commented:
/assign @xing-yang @NickrenREN @fengzixu

NickrenREN commented:
/triage accepted

xing-yang commented:
/ok-to-test

fengzixu commented:
/test pull-kubernetes-e2e-kind

fengzixu commented:
/test pull-kubernetes-e2e-gce-storage-slow

xing-yang commented on pkg/features/kube_features.go:
```diff
@@ -660,6 +660,12 @@ const (
+	// owner: @fengzixu
+	// alpha: v1.21
+	//
+	// Enables kubelet to detect CSI volume condition and send the event of the abnornal volume to the corresponding pod that is using it.
```
typo: abnornal volume -> abnormal volume condition

xing-yang commented on pkg/volume/csi/csi_client.go:
```diff
@@ -75,7 +75,7 @@ type csiClient interface {
 		volID string,
 		targetPath string,
-	) (*volume.Metrics, error)
+	) (*volume.Metrics, *csipbv1.NodeGetVolumeStatsResponse, error)
```
Need to add a NodeSupportsVolumeCondition.

xing-yang replied:
Instead of returning the entire NodeGetVolumeStatsResponse, can you just return VolumeCondition from NodeGetVolumeStatsResponse here?  Also whether to return nil or VolumeCondition should depends on the feature gate.

ehashman commented:
/priority important-soon

xing-yang commented:
/assign @msau42

msau42 commented:
/assign  @dashpole @Jiawei0227

Jiawei0227 commented on pkg/volume/csi/csi_client.go:
```diff
@@ -608,13 +610,19 @@ func (c *csiDriverClient) NodeGetVolumeStats(ctx context.Context, volID string,
 		VolumePath: targetPath,
 	}
 
+	var volumeCondition *csipbv1.VolumeCondition
```
I feel a little uncomfortable by having a new field being returned. I am thinking if it actually makes sense to add the volumeCondition field into the volume.Metrics type(and rename it to Status?).
So that 

1. we dont have to change a whole bunch of signature here. Or maybe we would because we need to change the name?
2. It will help us in the future if we want to add more fields to this.

fengzixu replied:
First of all, I have the same feeling with you
> I feel a little uncomfortable by having a new field being returned

I also contacted with @xing-yang about this approach IMO,

1. The current approach is the simplest way to reach our original target and move the responsibility of the external-health-monitor agent to kubelet by code freeze of v1.21 beta. And it's also the way which takes the smallest change for current kubelet implementation
2. The real issue is not in design of GetMetrics or MetricsProvider interface. The issue is we choose to re-use the volumeCondition field which is already gotten in GetMetrics function. If we want to have a more clear way to get volumeCondition, we can add new interface in pkg/volume/volume.go called StateProvider that contains a method called GetVolumeStats. And then add an argument using *csipbv1.VolumeUsage type to GetMetrics. After that we can remove the calling of NodeGetVolumeStats function inside GetMetrics function, call it out of side and pass the volumeUsage by argument.

```go
// Volume represents a directory used by pods or hosts on a node. All method
// implementations of methods in the volume interface must be idempotent.
type Volume interface {
	// GetPath returns the path to which the volume should be mounted for the
	// pod.
	GetPath() string

	// MetricsProvider embeds methods for exposing metrics (e.g.
	// used, available space).
	MetricsProvider
        
        // StateProvider
        StateProvider
}
```

type StateProvider interface {
	GetState() (*Metrics, *csipbv1.NodeVolumeStatsResp, error)
}

func (s *volumeStatCalculator) calcAndStoreStats() {
	// Find all Volumes for the Pod
	volumes, found := s.statsProvider.ListVolumesForPod(s.pod.UID)
	if !found {
		return
	}

	// Get volume specs for the pod - key'd by volume name
	volumesSpec := make(map[string]v1.Volume)
	for _, v := range s.pod.Spec.Volumes {
		volumesSpec[v.Name] = v
	}

	// Call GetMetrics on each Volume and copy the result to a new VolumeStats.FsStats
	var ephemeralStats []stats.VolumeStats
	var persistentStats []stats.VolumeStats
	for name, v := range volumes {
                 csiVolumeStats :=  v.GetState()
		metric, csiVolumeStats, err := v.GetMetrics(csiVolumeStats.GetUsage())
                 ......

                	if utilfeature.DefaultFeatureGate.Enabled(features.CSIVolumeHealth) {
			if csiVolumeStats.VolumeCondition.Abnormal {
				s.eventRecorder.Event(s.pod, v1.EventTypeWarning, "VolumeConditionAbnormal", csiVolumeStats.VolumeCondition.GetMessage())
			}
		}
              .....

fengzixu replied:
I prefer to add a new interface to implement the GetVolumeState function, other than modify the Metrics struct type. volume.Metrics should only include the metrics info

xing-yang replied:
We are calling one RPC NodeGetVolumeStats to get both metrics and volume condition, so we need one function that returns both results and then process them separately.

Jiawei0227 commented:
/assign

I only have one concern, otherwise it looks good to me.

xing-yang commented:
>  I am thinking if it actually makes sense to add the volumeCondition field into the volume.Metrics type(and rename Metric to Status or something?).

@dashpole can you comment on this?

dashpole commented on pkg/kubelet/server/stats/volume_stat_calculator.go:
```diff
@@ -128,6 +132,11 @@ func (s *volumeStatCalculator) calcAndStoreStats() {
 
+		if utilfeature.DefaultFeatureGate.Enabled(features.CSIVolumeHealth) {
+			if csiVolumeCondition.Abnormal {
+				s.eventRecorder.Event(s.pod, v1.EventTypeWarning, "VolumeConditionAbnormal", csiVolumeCondition.GetMessage())
```
Is there any way for someone to know _which_ volume is having problems, or is that included in the message?

xing-yang replied:
That's a good point.  GetMessage() shows the message from the storage system so it won't necessarily have volume name.  We should include volume name in the event here.

dashpole commented:
> I am thinking if it actually makes sense to add the volumeCondition field into the volume.Metrics type(and rename Metric to Status or something?).

 From an instrumentation perspective, I don't have a strong opinion on how you'd like to define the interface.  Based on the fact that most implementations return `nil`, i'd say it doesn't belong in the interface at all.  But given it is already part of the same CSI function, it seems like you probably have to add it.

xing-yang commented:
@msau42 Can you take a look?  We have a few options:
- implement it the current way as shown in the PR
- modify volume.Metrics to include volume condition
- modify volume.Metrics to include volume condition and rename it to volume.Stats, but that implies we should also rename GetMetrics to GetStats.  That would affect too many places.

fengzixu commented:
/test pull-kubernetes-node-e2e
--

fengzixu commented:
/test pull-kubernetes-node-e2e

fengzixu commented:
/test pull-kubernetes-unit

fengzixu commented:
/retest

fengzixu commented:
/retest

fengzixu commented:
/retest

xing-yang commented on pkg/kubelet/server/stats/volume_stat_calculator.go:
```diff
@@ -122,13 +127,18 @@ func (s *volumeStatCalculator) calcAndStoreStats() {
 
+		if utilfeature.DefaultFeatureGate.Enabled(features.CSIVolumeHealth) {
+			if volumeStats.Abnormal != nil && volumeStats.Message != nil && (*volumeStats.Abnormal) {
+				s.eventRecorder.Event(s.pod, v1.EventTypeWarning, "VolumeConditionAbnormal", fmt.Sprintf("%s: %s", name, *volumeStats.Message))
```
fmt.Sprintf("Volume %s: %s", name, *volumeStats.Message

xing-yang commented on pkg/kubelet/stats/host_stats_provider_fake.go:
```diff
@@ -69,18 +69,18 @@ func (f *fakeHostStatsProvider) getPodLogStats(podNamespace, podName string, pod
 }
 
-func fakeMetricsProvidersToStats(metricsProviders []volume.MetricsProvider, rootFsInfo *cadvisorapiv2.FsInfo) (*statsapi.FsStats, error) {
+func fakeMetricsProvidersToStats(metricsProviders []volume.StatsProvider, rootFsInfo *cadvisorapiv2.FsInfo) (*statsapi.FsStats, error) {
```
metricsProviders -> statsProviders

xing-yang commented on pkg/volume/azuredd/azure_common.go:
```diff
@@ -88,17 +88,17 @@ func makeGlobalPDPath(host volume.VolumeHost, diskURI string, isManaged bool) (s
 
 func makeDataDisk(volumeName string, podUID types.UID, diskName string, host volume.VolumeHost, plugin *azureDataDiskPlugin) *dataDisk {
-	var metricProvider volume.MetricsProvider
+	var metricProvider volume.StatsProvider
```
metricProvider -> statsProvider

xing-yang commented on pkg/volume/csi/csi_client_test.go:
```diff
@@ -674,13 +705,108 @@ type VolumeStatsOptions struct {
-func TestVolumeStats(t *testing.T) {
+func TestVolumeHealthEnable(t *testing.T) {
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.CSIVolumeHealth, true)()
+	spec := volume.NewSpecFromPersistentVolume(makeTestPV("test-pv", 10, "metrics", "test-vol"), false)
```
metrics -> stats

fengzixu commented:
Because we rename the interface name from MetricsProver to StatsProvider and rename Metrics struct to Stats struct. This change will cause lots of modifications about "name".

After I talked with @xing-yang. I think have two options

1. Correct all names about "metrics" to "stats" in this PR. The disadvantage is we will miss the deadline of codefreeze
2. Make sure "sending events of abnormal volume in kubelet" feature is merged first. For "name" issue, I can submit another PR to fix them.

Because the codefreeze date is 9th Mar. I need your opinion about the scope of this PR ASAP @msau42 @dashpole @xing-yang

fengzixu commented:
/retest

fengzixu commented:
/retest

fengzixu commented:
/retest

fengzixu commented:
/retest

fengzixu commented:
/retest

xing-yang commented:
CC @dchen1107

fengzixu commented:
/retest

fengzixu commented:
/retest

fengzixu commented:
/retest

fengzixu commented:
/retest

xing-yang commented:
/retest

xing-yang commented:
/retest

xing-yang commented:
/retest

Jiawei0227 reviewed (commented):
This looks good. I think just need a rebase and we can get it merged

msau42 commented on pkg/volume/csi/csi_client.go:
```diff
@@ -612,38 +614,75 @@ func (c *csiDriverClient) NodeGetVolumeStats(ctx context.Context, volID string,
+		return nil, err
+	}
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.CSIVolumeHealth) && isSupportNodeVolumeCondition {
```
Can you move the feature gate check before `supportNodeGetVolumeCondition`?

msau42 commented on pkg/volume/csi/csi_metrics_test.go:
```diff
@@ -21,23 +21,38 @@ import (
+			volumeConditionSet: true,
+		},
+		{
+			name:               "with valid name and volume id",
```
Can you update the name to add that the capability is disbled?

msau42 commented on pkg/volume/volume.go:
```diff
@@ -92,6 +92,15 @@ type Metrics struct {
+
+	// Normal volumes are available for use and operating optimally.
+	// An abnormal volume does not meet these criteria.
+	// This field is REQUIRED.
```
I guess it's required in the csi spec if you support the capability, but from a k8s volume pov, it's an optional field that's only filled in by some csi drivers right?

msau42 commented:
Change mostly lgtm, just some minor wording suggestions.

We'll need approval from sig-node for the changes to the kubelet metric processing.

msau42 commented:
/approve

xing-yang commented:
/assign @dchen1107 
/assign @ehashman

Can someone from SIG-Node please help review this?  This is targeting 1.21.

xing-yang commented:
/assign @random-liu

xing-yang commented:
/assign @sjenning

sjenning commented:
This was hard to review for me because there is so much noise in the diff due to adding the `EventRecorder` and renaming `stat` to `metrics`.  It is pretty difficult to extract the actual KEP related functional change from this PR.

Please break this out into 2-3 commits which isolates the actual KEP work from the non-functional changes.

Maybe:
* rename stat to metrics
* add EventRecorder to ResourceAnalyzer
* add CSIVolumeHealth feature

fengzixu commented:
@sjenning You can check it now

xing-yang commented on pkg/volume/csi/csi_client_test.go:
```diff
@@ -74,38 +86,53 @@ func (c *fakeCsiDriverClient) NodeGetInfo(ctx context.Context) (
 
 func (c *fakeCsiDriverClient) NodeGetVolumeStats(ctx context.Context, volID string, targetPath string) (
-	usageCountMap *volume.Metrics, err error) {
+	usageCountMap *volume.Stats, err error) {
```
there's still renaming in the first commit

fengzixu committed:
feature: add CSIVolumeHealth feature and gate

1. add EventRecorder to ResourceAnalyzer
2. add CSIVolumeHealth feature and gate

sjenning commented:
/approve

msau42 commented:
/milestone v1.21

xing-yang commented:
/test pull-kubernetes-integration

xing-yang commented:
/test pull-kubernetes-integration

xing-yang commented:
/test pull-kubernetes-integration

xing-yang commented:
Test failure is not related to this change.

/lgtm

xing-yang commented:
@fengzixu Please rename the PR to: `feature: add CSIVolumeHealth feature and gate`
